workflow/controller/cache/cache.go

```diff
@@ -0,0 +1,124 @@
+package cache
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"sync"
+
+	apierr "k8s.io/apimachinery/pkg/api/errors"
+
+	"github.com/argoproj/argo/errors"
+
+	log "github.com/sirupsen/logrus"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	wfv1 "github.com/argoproj/argo/pkg/apis/workflow/v1alpha1"
+)
+
+var cacheKeyRegex = regexp.MustCompile("^[a-zA-Z0-9][-a-zA-Z0-9]*$")
+
+type MemoizationCache interface {
+	Load(key string, configMapName string) (*CacheEntry, error)
+	Save(key string, nodeId string, value *wfv1.Outputs, configMapName string) error
+}
+
+type CacheEntry struct {
+	NodeID  string       `json:"nodeID"`
+	Outputs wfv1.Outputs `json:"outputs"`
+	// TODO: Add creation timestamp
```

Comment:
I think we need a date here to indicate expiry

Reply:
discussed with @jessesuen and @simster7 and decided this will be in a fast follow